core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java

```diff
@@ -19,50 +19,120 @@
+            if (b1 == LUCENE_CODEC_HEADER_BYTE) {
+                // Read 3 more bytes, meaning a whole integer has been read
+                byte b2 = headerStream.readByte();
+                byte b3 = headerStream.readByte();
+                byte b4 = headerStream.readByte();
+                // Convert the 4 bytes that were read into an integer
+                int header = ((b1 & 0xFF) << 24) + ((b2 & 0xFF) << 16) + ((b3 & 0xFF) << 8) + ((b4 & 0xFF) << 0);
+                // We confirm CodecUtil's CODEC_MAGIC number (0x3FD76C17)
+                // ourselves here, because it allows us to read the first
+                // byte separately
+                if (header != CodecUtil.CODEC_MAGIC) {
+                    throw new TranslogCorruptedException("translog looks like version 1 or later, but has corrupted header");
+                }
+                // Confirm the rest of the header using CodecUtil, extracting
+                // the translog version
+                int version = CodecUtil.checkHeaderNoMagic(new InputStreamDataInput(headerStream), TranslogWriter.TRANSLOG_CODEC, 1, Integer.MAX_VALUE);
+                switch (version) {
+                    case TranslogWriter.VERSION_CHECKSUMS:
+                        throw new IllegalStateException("pre-2.0 translog found [" + path + "]");
+                    case TranslogWriter.VERSION_CHECKPOINTS:
+                        assert path.getFileName().toString().endsWith(Translog.TRANSLOG_FILE_SUFFIX) : "new file ends with old suffix: " + path;
+                        assert checkpoint.numOps >= 0 : "expected at least 0 operatin but got: " + checkpoint.numOps;
+                        assert checkpoint.offset <= channel.size() : "checkpoint is inconsistent with channel length: " + channel.size() + " " + checkpoint;
+                        int len = headerStream.readInt();
+                        if (len > channel.size()) {
+                            throw new TranslogCorruptedException("uuid length can't be larger than the translog");
+                        }
+                        BytesRef ref = new BytesRef(len);
+                        ref.length = len;
+                        headerStream.read(ref.bytes, ref.offset, ref.length);
+                        BytesRef uuidBytes = new BytesRef(translogUUID);
+                        if (uuidBytes.bytesEquals(ref) == false) {
+                            throw new TranslogCorruptedException("expected shard UUID [" + uuidBytes + "] but got: [" + ref + "] this translog file belongs to a different translog");
+                        }
+                        return new ImmutableTranslogReader(checkpoint.generation, channel, path, ref.length + CodecUtil.headerLength(TranslogWriter.TRANSLOG_CODEC) + RamUsageEstimator.NUM_BYTES_INT, checkpoint.offset, checkpoint.numOps);
+                    default:
+                        throw new TranslogCorruptedException("No known translog stream version: " + version + " path:" + path);
```

Comment:
It's great that this exception message includes the path ... can we fix the others to do so as well?